src/crimson/os/seastore/journal.cc

```diff
@@ -0,0 +1,365 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include <boost/iterator/counting_iterator.hpp>
+
+#include "crimson/os/seastore/journal.h"
+
+#include "include/intarith.h"
+#include "crimson/os/seastore/segment_manager.h"
+
+namespace {
+  seastar::logger& logger() {
+    return crimson::get_logger(ceph_subsys_filestore);
+  }
+}
+
+namespace crimson::os::seastore {
+
+Journal::Journal(SegmentManager &segment_manager)
+  : block_size(segment_manager.get_block_size()),
+    max_record_length(
+      segment_manager.get_segment_size() -
+      p2align(ceph::encoded_sizeof_bounded<segment_header_t>(),
+	      size_t(block_size))),
+    segment_manager(segment_manager) {}
+
+
+Journal::initialize_segment_ertr::future<> Journal::initialize_segment(
+  Segment &segment)
+{
+  logger().debug("initialize_segment {}", segment.get_segment_id());
+  // write out header
+  ceph_assert(segment.get_write_ptr() == 0);
+  bufferlist bl;
+  auto header = segment_header_t{
+    current_journal_segment_seq++,
+    segment.get_segment_id(),
+    current_replay_point};
+  ::encode(header, bl);
+
+  written_to = segment_manager.get_block_size();
+  return segment.write(0, bl).handle_error(
+    init_ertr::pass_further{},
+    crimson::ct_error::all_same_way([] { ceph_assert(0 == "TODO"); }));
+}
+
+ceph::bufferlist Journal::encode_record(
+  segment_off_t mdlength,
+  segment_off_t dlength,
+  record_t &&record)
+{
+  bufferlist metadatabl;
+  record_header_t header{
+    mdlength,
+    dlength,
+    current_journal_seq,
+    0 /* checksum, TODO */,
+    record.deltas.size(),
+    record.extents.size()
+  };
+  ::encode(header, metadatabl);
+  for (const auto &i: record.deltas) {
+    ::encode(i, metadatabl);
+  }
+  bufferlist databl;
+  for (auto &i: record.extents) {
+    databl.claim_append(i.bl);
+  }
+  if (metadatabl.length() % block_size != 0) {
+    metadatabl.append(
+      ceph::bufferptr(
+	block_size - (metadatabl.length() % block_size)));
+  }
+  ceph_assert(metadatabl.length() == (size_t)mdlength);
```

Comment:
is this still true after the padding in line 70?

Reply:
Yeah, get_encoded_record_length already accounts for the padding, the assert is validating/documenting that predicted it correctly.  It's a bit confusing because the encoding includes the encoded metadata and data sizes.